bloxid/v0.go

```diff
@@ -48,16 +50,78 @@ var (
 
 // NewV0 parse a string into a typed guid, return an error
 // if the string fails validation.
+/*
 func NewV0(bloxid string) (*V0, error) {
 	return parseV0(bloxid)
 }
+*/
 
 const (
 	V0Delimiter   = "."
 	bloxidTypeLen = 4
 )
 
-func parseV0(bloxid string) (*V0, error) {
+type EncodeDecodeOpts func(o *V0Options)
+type NewV0Opts func(fnOpts ...EncodeDecodeOpts) *V0Options
+
+// NewV0 parse a string into a typed guid, return an error
+// if the string fails validation.
+
+func NewV0(bloxid string, opts ...*V0Options) (*V0, error) {
```

Comment:
```suggestion
func NewV0(bloxid string, opts ...EncodeDecodeOpts) (*V0, error) {
```

```
v0, err := NewV0("", 
    WithEntityType(test.entityType),
    WithRealm(test.realm),
    WithHashID(test.intID),
    WithHashIDSalt(test.salt)
)
```

```
v0, err := NewV0("blox0.infra.host.us-com-1.zdud52youveke5sovyoc66cjxw3l55jc", 
    WithHashIDSalt(test.salt)
)
```

Reply:
I feel using NewV0 to both generation and parsing would be an issue if someone passes empty id by mistake (to parse the id). The caller expects error but NewV0 returns random id. @drewwells @wplu

Reply:
the person is welcome to check the Scheme() to ensure it is SchemeHashID = 'hashid'

Reply:
the person is also welcome to check that the parsed object is the same as the original string:
```
v0, err := NewV0("", 
    WithEntityDomain("infra"),
    WithEntityType("group"),
    WithRealm("us-com-1"),
)
// v0.String() -> "blox0.infra.group.us-com-1.fsr4beqqkdhpirlegenwsjyqhj4pnm3h"
parsed, err := NewV0(""  /* bug: did not pass v0 */)
assert(v0.String() == parsed.String())
```